src/orange/orangeinp/RevolvedPolygon.cc

```diff
@@ -0,0 +1,273 @@
+/*!
+ * Recursively construct convex regions, subtracting out concavities.
+ */
+NodeId RevolvedPolygon::make_levels(detail::VolumeBuilder& vb,
+                                    VecReal2 const& polygon,
+                                    RevolvedPolygon::SubIndex si) const
+{
+    // Find the convex hull
+    detail::ConvexHullFinder<real_type> hull_finder(polygon, vb.tol());
+    auto convex_hull = hull_finder.make_convex_hull();
+    auto concave_regions = hull_finder.calc_concave_regions();
+
+    // Build the convex region
+    auto filtered_convex_hull
+        = detail::filter_collinear_points(convex_hull, vb.tol().abs);
+    NodeId result = this->make_region(vb, filtered_convex_hull, si);
+
+    // Return early if there are no concave regions to process
+    if (concave_regions.empty())
+    {
+        return result;
+    }
+
+    // Create a vector of all concave regions, via recursion
+    std::vector<NodeId> concave_nodes(concave_regions.size());
+    for (auto i : range(static_cast<size_type>(concave_nodes.size())))
+    {
+        concave_nodes[i] = this->make_levels(
+            vb, concave_regions[i], SubIndex{si.level + 1, i});
+    }
+
+    auto level_ext = this->make_level_ext(si);
+
+    // Create a union of all concave regions
+    NodeId concave_union
+        = vb.insert_region(Label{label_, level_ext + ".cu"},
+                           Joined{op_or, std::move(concave_nodes)});
+
+    // Create a negation of this union
+    auto sub_node = vb.insert_region(Label{label_, level_ext + ".ncu"},
+                                     Negated{concave_union});
+
+    // Subtract concave regions from the convex hull
+    return vb.insert_region(Label{label_, level_ext + ".d"},
+                            Joined{op_and, {result, sub_node}});
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Revolved a convex polygon around the \em z axis.
+ *
+ * The polygon be strictly counterclockwise, i.e., it must not contain any
+ * collinear points. Construction is done by creating a set of outer subregions
+ * that define the outer edge of the polygon and subtracting out a set of inner
+ * subregions.
+ */
+NodeId RevolvedPolygon::make_region(detail::VolumeBuilder& vb,
+                                    VecReal2 const& polygon,
+                                    SubIndex si) const
+{
+    // The polygon should have a strictly clockwise orientation
+    CELER_VALIDATE(has_orientation(make_span(polygon),
+                                   detail::Orientation::counterclockwise),
+                   << "polygon must be specified in strictly counterclockwise "
+                      "order");
+
+    SoftEqual<real_type> soft_equal(vb.tol().rel, vb.tol().abs);
+    SoftZero<real_type> soft_zero(vb.tol().abs);
+
+    auto n = polygon.size();
+    std::vector<NodeId> outer_nodes;
+    std::vector<NodeId> inner_nodes;
+
+    auto next_idx = [n](size_type i) { return (i + 1 < n) ? i + 1 : 0; };
```

Comment:
I think the intent matches the semantics of modulo (wrapping around) and improves readability:
```suggestion
    auto next_idx = [n](size_type i) { return i % n; };
```
or since this is only used in one place, you can put it inline there.

Reply:
I only did it this way on the basis of comments on a previous MR. Switched to modulus.